### PR TITLE
Fixed operator[] const in UnorderedMap* templates

### PR DESCRIPTION
**Pull request by fuchsto**
Addresses #259

### CODE REVIEW

File: dash/include/dash/map/UnorderedMap.h [hunk 1]
```diff
@@ -513,17 +513,42 @@ class UnorderedMap
...
+    DASH_LOG_TRACE("UnorderedMap.at", "byte offset of mapped member:",
+                   mapped_offs);
+    // Increment pointers to element by byte offset of mapped value member:
+    if (lptr_value != nullptr) {
+      // Convert to char pointer for byte-wise increment:
+      char * b_lptr_mapped  = reinterpret_cast<char *>(lptr_value);
+      b_lptr_mapped        += mapped_offs;
+      // Convert to mapped type pointer:
+      lptr_mapped           = reinterpret_cast<mapped_type *>(b_lptr_mapped);
+    }
```
Comment (fmoessbauer) [inline dash/include/dash/map/UnorderedMap.h]: This looks way to complex to me. Why can't you just use `lptr_mapped = &(lptr_value->second)`?

Comment (fuchsto) [reply dash/include/dash/map/UnorderedMap.h]: If I remember correctly, this became necessary for some MPI runtimes.

The actual reasons behind this are that 
- `std::pair` cannot be used as MPI data type directly.
- Offset-to-member only works reliably with `offsetof` in the general case

I have to use `offsetof` in line 535 above as there is no instance of `value_type` available that could be used to calculate the member offset as `l_ptr_value` is possibly undefined.
Using `std::declval()` instead (to generate a compile-time pseudo-instance for member resolution) only works if `Key` and `Mapped` are default-constructible.

Finally, the distance obtained from `&(lptr_value->second) - lptr_value` had different alignment than the address obtained via `offsetof` in some cases, depending on the combination of MPI runtime and compiler. Apparently some compilers / standard libs have special treatment (padding?) for members of `std::pair` such that `__builtin_offsetof(type, member)` differs from the member-offset provided by the type system.

The alternative, using `offsetof` (which resolves to `__builtin_offsetof` automatically if needed) and manual pointer increments works, however.

Comment (devreal) [reply dash/include/dash/map/UnorderedMap.h]: Why is `&(lptr_value->second)` valid in cases where `value_type` is not of `standard_layout`?

**Comment by fuchsto** [comment]
@fmoessbauer Bounce for review of my recent changes

**Review by fmoessbauer — APPROVED** [review]
Looks good! However the pointer arithmetics on `lptr_value` are still not clear to me. Let's discuss this tomorrow.

**Comment by devreal** [comment]
Please thoroughly document the `lptr_value` quirk before merging.

**Comment by fuchsto** [comment]
@devreal 
I moved the quirky bits to a separate private method and added documentation there.

**Comment by fuchsto** [comment]
Aaah ... this is embarassing ... I forgot to `git add` my recent changes. Will create a new pull request, run CI, and merge it.

**Pull request merged**